content/updates/understanding-the-terraform-check-block-feature.md

```diff
@@ -0,0 +1,207 @@
+    error_message = <<EOF
+    Device ${data.tailscale_device.default.name} is not tagged with the correct tags.
+    The list of expected tags is: [${join(",", local.tailscale_tags)}]
+    EOF
+  }
+}
+```
+
+Example of the failed assertion:
+```sh
+Warning: Check block assertion failed
+  on ../../../terraform-aws-tailscale/checks.tf line 19, in check "device":
+  19:     condition     =  sort(data.tailscale_device.default.tags) == sort(tolist(local.tailscale_tags))
+    ├────────────────
+    │ data.tailscale_device.default.tags is set of string with 1 element
+    │ local.tailscale_tags is tuple with 1 element
+Device mp-automation-tailscale-subnet-router.cat-crocodile.ts.net is not tagged with the correct tags.
+```
+
+### Limitations to consider
+
+1. **Approval is required for `terraform apply`**
+
+In case of a successful assertion, Terraform requires approval for every `apply` operation due to a configuration reload needed to verify a check block, e.g.:
+
+```sh
+Terraform will perform the following actions:
+
+  # module.tailscale_subnet_router.data.tailscale_device.default will be read during apply
+  # (config will be reloaded to verify a check block)
+ <= data "tailscale_device" "default" {
+      + addresses = [
+          + "100.100.48.62",
+        ]
+      + id        = "8181818181818181"
+      + name      = "mp-automation-tailscale-subnet-router.cat-crocodile.ts.net"
+      + tags      = [
+          + "tag:mp-automation-tailscale-subnet-router",
+        ]
+      + wait_for  = "30s"
+    }
+
+Plan: 0 to add, 0 to change, 0 to destroy.
+
+Do you want to perform these actions?
+  Terraform will perform the actions described above.
+  Only 'yes' will be accepted to approve.
+
+  Enter a value:
+```
+
+This should be considered when introducing check to existing CI/CD pipelines.
+
+2. **Multiple data resource blocks are not supported**
+
+Unfortunately, it's possible to define only one data source per check at the moment. Otherwise, the error will be thrown:
+
+```sh
+This check block already has a data resource defined at main.tf:83,3-27.
+```
+
+This limitation restrains creating complex assertions within one `check` block.
+
+## Are there any potential pitfalls?
+
+In addition to the limitations above, we recommend paying attention to a couple of things:
+
+* While, by design, Terraform should not halt due to a check, the use of a data source can increase the operation's execution time and might cause timeout errors if Terraform fails to fetch it. Consider setting a retry limit if the provider offers this option.
+* As `terraform plan` and `terraform apply` represent different stages in the workflow, the purpose of checks can also diverge into post-plan, post-apply, and the ones relevant for both cases. We see great potential for improvement here, such as the possibility of labeling or ignoring checks for a particular stage so that checks could be built-in as smoothly as possible.
+
```

Comment:
@gberenice I'm thinking we should add another section here. Something like... 

```suggestion
## Our Thoughts

Overall, we're slightly disappointed and puzzled by this new feature. When initially hearing about it, our team was excited and we were eager to try it out. But when actually using it and working on ways that we could work it into our terraform code, we found that it wasn't actually as useful as we would like. It does not solve our needs around unit testing individual resources, it is attached to the lifecycle of `plan` and `apply`, and we don't have a means by which to flag a `check` block as a critical assertion that should halt execution. 

We believe there is still a lot to be done on the terraform testing and validation front and we're eager to see more functionality from the `check` block and language as a whole. 
```

We'd need to reword the conclusion a bit, but what do you think?  I want us to be honest... because I'm not very excited about this feature anymore and I think I'm still struggling on where we can use it.

Reply:
Haha, is that your impression after reading this draft or you tried to build-in the `check` feature in your work?
The conclusion is on its way 👍

Reply:
Final conclusion was merged with "Our thoughts" block.